### DISCUSSION TITLE
Results.Distinct dosent actually show the total reactions collected

### DISCUSSION THREAD

**Discussion by samjesus8** [DSharpPlus/DSharpPlus#1393] [Support] [answered]
# Summary
I have this command called >mid which is basically like an Anime Girl voting system, it displays a girl from a JSON file and people have 20 seconds to vote. Once the voting period is over it collects the reactions and displays them in an embed. The problem I am having is that it doesn't actually add up the totals and instead displays each result separately which is weird.

# Details
The way the command works is that it makes a "Card" based on a JSON file which has the name, anime name, and the image URL of the character and I put this in an array. All this is chosen at random obviously. I then take each bit of the array and display it in a simple embed with the custom emojis I loaded in at the start of the code.

I then collect the reactions and use .Distinct() to only show the user reactions apart from the bot itself. When an emote has more than 2 reactions, in the total it doesn't display the total as 2, instead it displays the same thing twice with a total of 1. So how do I make it that it collects the reactions of both emotes and actually adds up the totals???

This is the code and I will provide a SS of the output in a discord server where multiple reactions have been made on the same emote:

```
        [Command("mid")]
        [Cooldown(10, 300, CooldownBucketType.User)]
        public async Task MidOrNotMid(CommandContext ctx)
        {
            DiscordEmoji[] emojiOptions = { DiscordGuildEmoji.FromName(ctx.Client, ":MID_EMOTE:", true), DiscordGuildEmoji.FromName(ctx.Client, ":NOT_MID_EMOTE:", true) }; //Declaring Emotes from a custom server
            var interactivity = ctx.Client.GetInteractivity();
            var random = new Random();
            var duration = TimeSpan.FromSeconds(20); //Declaring the timespan for the command

            List<String[]> messages = new List<String[]>(); //[0] = Name, [1] = Anime Name, [2] = Image URL

            var waifuCard = new WaifuCardBuilder(); //Custom made class for the building of this card

            string[] card = { waifuCard.Name, waifuCard.AnimeName, waifuCard.ImageURL };
            messages.Add(card); //Making the character card to display

            int index = random.Next(messages.Count);

            var testEmbed = new DiscordMessageBuilder()
                .AddEmbed(
                new DiscordEmbedBuilder()
                .WithAuthor("MID OR NOT MID?? | Cast your votes below | You have 20 seconds to vote")
                .WithTitle("***" + messages[index][0] + "***") //Gets the Name of the character
                .WithDescription(messages[index][1]) //Gets the Anime name it came from
                .WithImageUrl(messages[index][2]) //Gets the ImageURL to show a photo
                .WithColor(DiscordColor.Blue)
                .WithFooter("Use '>mid info' to view the instructions on this command")
                );
            var pollMSG = await ctx.Channel.SendMessageAsync(testEmbed); //Storing the embed in a variable

            foreach (var option in emojiOptions)
            {
                await pollMSG.CreateReactionAsync(option); //Adding reactions to the message
            }

            var result = await interactivity.CollectReactionsAsync(pollMSG, duration); //Collects reactions after 20 seconds
            var dResult = result.Distinct(); //Gets only the user reactions
            var results = dResult.Select(x => $"{x.Emoji}: {x.Total}"); //Sorts the results

            var resultsEmbed = new DiscordMessageBuilder()
                .AddEmbed(
                new DiscordEmbedBuilder()
                .WithTitle("***Results***")
                .WithColor(DiscordColor.Azure)
                .WithDescription(string.Join("\n", results))
                );
            await ctx.Channel.SendMessageAsync(resultsEmbed); //Display results in an embed 20 seconds later
        }
```

This is what happened when it was used in a public server:

![image](https://user-images.githubusercontent.com/98812930/192844911-0c7b284f-ae7d-48cd-ba75-cd39224cf1ee.png)

It shows a total of 1, 4 times which makes no sense. It should be that emote and then 4 as the total (Distinct works since it ignores the bot reaction)

**Comment by akiraveliara** [comment]
from what i can tell, this is working as intended. `Distinct` isn't needed to ignore the bot reactions, that's taken care of by `CollectReactionsAsync`, instead, it returns a list of all different `Reaction` objects. since these four all have a different user who created them, they'll all count as distinct.

you can work around this by
- counting different reactions yourself, in a foreach loop for example
- using `GroupBy` and  `Count` instead of `Distinct`

edit: checking again, this may not be working as intended, but the workarounds should still work. i'll look into the issue closer.

**Comment by akiraveliara** [comment, marked as answer]
The underlying issue has now been fixed. You won't need that `Distinct` call, it won't achieve what you want it to or what your comment claims to achieve, but it will work as of the latest nightly version.

**Reply by samjesus8** [reply]
Ok thanks so much, I will try your suggestions and see if i can get something working

**Comment by samjesus8** [comment]
Yeah thank you ever so much man, this took me weeks of searching to find an answer. For anyone else that is struggling with this type of issue, I will post my solution here. I basically used a for loop as suggested and have 2 if statements to check which emoji is which and have a count based on that. Once the loop is done, your counts are your totals so you can just display that in your message

```
            var result = await interactivity.CollectReactionsAsync(pollMSG, duration); //My Original message i was trying to get reactions from

            int midCount = 0; //Declaring 2 variables for both emojis. This also ignores other emojis you don't want to count
            int notMidCount = 0;

            foreach (var emoji in result) 
            {
                if (emoji.Emoji == emojiOptions[0])  //If the emoji is equal to my MID emoji, increment the mid count by 1
                {
                    midCount++;
                }
                if (emoji.Emoji == emojiOptions[1]) //If the emoji is equal to my NOT_MID emoji, increment the notmid count by 1
                {
                    notMidCount++;
                }
            }
//At this point in the code you can do a simple Console.WriteLine() to check your counts and see if they are correct
            var resultsEmbed = new DiscordMessageBuilder()
                .AddEmbed(
                new DiscordEmbedBuilder()
                .WithTitle("***Results***")
                .WithColor(DiscordColor.Azure)
                .WithDescription(emojiOptions[0] + ": " + midCount + "\n" +
                                 emojiOptions[1] + ": " + notMidCount) 
                );
            await ctx.Channel.SendMessageAsync(resultsEmbed);
```

Here is my test that I did on a Discord Server and it works like a charm:

![image](https://user-images.githubusercontent.com/98812930/193275914-aa72d642-5297-4846-a650-010ec0e6f45f.png)
